File: julia/ch1/Ex124.jl
"""
```jldoctest
julia> include("c:/Users/Yohan/IdeaProjects/sicp/julia/ch1/ExXXX.jl")

julia> ExXXX.func(arg)
```
"""
module Ex124

fast_times = 3

sq(n) = n * n

divides(a, b) = b % a == 0

function find_divisor(n, test)
    if sq(test) > n
        return n
    elseif divides(test, n)
        return test
    else
        return find_divisor(n, test + 1)
    end
end

smallest_divisor(n) = find_divisor(n, 2)

# prime(n) = n == smallest_divisor(n)
prime(n) = fast_prime(n, fast_times)

function fast_prime(n, times)
    if times == 0
        return true
    elseif fermat_test(n)
        return fast_prime(n, times-1)
    else
        return false
    end
end 

function expmod(base, exp, m)
    if exp == 0
        return 1
    elseif divides(2, exp)
        return rem(sq(expmod(base, div(exp,2), m)), m)
    else
        return rem(base * expmod(base, exp-1, m), m)
    end
end

function fermat_test(n)
    try_it(a) = expmod(a, n, n) == a
    return try_it(1 + random(n - 1))
end

random(limit) = rem(abs(rand(Int, 1)[1]), limit)

function report_prime(n)
    if prime(n)
        println("$n ***")
    else
        println("$n")
    end
end

function search_helper(n, finish)
    if n <= finish
        report_prime(n)
        search_helper(n + 2, finish)
    else
        println("finished")
    end
end

search_for_primes(start, finish) = 
    search_helper(divides(2, start) ? start + 1 : start, finish)


end

File: erlang/ch1/ex124.erl
%%%-------------------------------------------------------------------
%%% @author Yohan Lee
%%% @copyright (C) 2019
%%% @doc
%%% cd("c:/Users/Yohan/IdeaProjects/sicp/erlang/ch1").
%%% compile with > c("exXXX.erl").
%%% if already compiled, load with > l(exXXX).
%%% run with > exXXX:main().
%%%
%%% @end
%%%-------------------------------------------------------------------
-module(ex124).
-author("Yohan").

%% API
-export([main/0]).

main() ->
  io:fwrite("~w\n", [search_for_primes(1000, 1020)]),
  io:fwrite("~w\n", [search_for_primes(10000, 10040)]),
  io:fwrite("~w\n", [search_for_primes(100000, 100080)]).


smallest_divisor(N) -> find_divisor(N, 2).

find_divisor(N, Test) when Test * Test > N -> N;
find_divisor(N, Test) ->
  if
    N rem Test == 0 -> Test;
    true -> find_divisor(N, Test + 1)
  end.

%%prime(N) -> N == smallest_divisor(N).
prime(N) -> fast_prime(N, 3).

fast_prime(_, 0) -> true;
fast_prime(N, Times) ->
  case fermat_test(N) of
    true -> fast_prime(N, Times-1);
    false -> false
  end.

square(N) -> N * N.

random(Limit) -> rand:uniform(Limit) - 1.

even(N) -> N rem 2 == 0.

expmod(_, 0, _) -> 1;
expmod(Base, Exp, M) ->
  case even(N) of
    true -> square(expmod(Base, Exp div 2, M)) rem M;
    false -> Base * expmod(Base, Exp - 1, M) rem M
  end.

fermat_test(N) ->
  TryIt = fun (A) -> expmod(A, N, N) == A end,
  TryIt(1 + random(N - 1)).

report_prime(N, Prime) ->
  if
    Prime -> io:fwrite("~w ***\n", [N]);
    true -> io:fwrite("~w \n", [N])
  end.

prime_test(N) -> report_prime(N, prime(N)).

search_helper(N, End) ->
  if
    N =< End -> prime_test(N), search_helper(N + 2, End);
    true -> io:fwrite("finished\n")
  end.

search_for_primes(Start, End) ->
  if
    Start rem 2 == 1 -> search_helper(Start, End);
    true-> search_helper(Start + 1, End)
  end.
